polkadot/node/core/prospective-parachains/src/fragment_tree.rs

```diff
@@ -763,43 +763,106 @@ impl FragmentTree {
 	/// The intention of the `required_path` is to allow queries on the basis of
 	/// one or more candidates which were previously pending availability becoming
 	/// available and opening up more room on the core.
-	pub(crate) fn select_child(
+	pub(crate) fn select_children(
 		&self,
 		required_path: &[CandidateHash],
+		count: u32,
 		pred: impl Fn(&CandidateHash) -> bool,
-	) -> Option<CandidateHash> {
+	) -> Vec<CandidateHash> {
 		let base_node = {
 			// traverse the required path.
 			let mut node = NodePointer::Root;
 			for required_step in required_path {
-				node = self.node_candidate_child(node, &required_step)?;
+				if let Some(next_node) = self.node_candidate_child(node, &required_step) {
+					node = next_node;
+				} else {
+					return vec![]
+				};
 			}
 
 			node
 		};
 
-		// TODO [now]: taking the first selection might introduce bias
+		// TODO: taking the first best selection might introduce bias
 		// or become gameable.
 		//
 		// For plausibly unique parachains, this shouldn't matter much.
 		// figure out alternative selection criteria?
-		match base_node {
+		self.select_children_inner(base_node, count, count, &pred, &mut vec![])
+	}
+
+	// Try finding a candidate chain starting from `base_node` of length `expected_count`.
+	// If not possible, return the longest we could find.
+	// Does a depth-first search, since we're optimistic that there won't be more than one such
+	// chains. Assumes that the chain must be acyclic.
```

Comment:
What happens if they are cyclic? What is the worst case complexity here, assuming parachains misbehave as much as they possibly can?

Reply:
that's a good question. we stop the search on the respective branch when we see that node A has a child that we've already visited. maybe we should just return an error. I don't think it's a valid state transition to accept. Is it?

Even if we were to accept cycles, we'd still be bounded by a function of the requested number of candidates.

Reply:
I modified the code in the latest revision to accept cycles in the fragment trees. This is actually a documented behaviour of prospective-parachains even with some tests for it. See: https://github.com/paritytech/polkadot-sdk/blob/74b597fcaf143d8dd7f8d40e59f51065514f21d7/polkadot/node/core/prospective-parachains/src/fragment_tree.rs#L43
I also documented the performance considerations. Cycles are not an issue because we are limited by the core count anyway so we won't loop indefinitely